SCAutolib/models/gui.py

```diff
@@ -0,0 +1,355 @@
+            kwargs_list = ["=".join((key, repr(value)))
+                           for key, value in kwargs.items()]
+            args_list = [repr(value) for value in list(args[1:])]
+            all_args = ", ".join(args_list + kwargs_list)
+            logger.info(f'Calling GUI.{func.__name__}({all_args})')
+            func(*args, **kwargs)
+        return wrapper
+
+    @action_decorator
+    @log_decorator
+    def click_on(self, key: str, timeout: float = 30):
+        """Clicks matching word on the screen.
+
+        """
+        logger.info(f"Trying to find key='{key}' to click on.")
+
+        end_time = time() + timeout
+        item = None
+        first_scr = None
+
+        # Repeat screenshotting, until the key is found
+        while time() < end_time:
+            # Capture the screenshot
+            screenshot = self.screen.screenshot()
+
+            last_scr = screenshot
+            if first_scr is None:
+                first_scr = screenshot
+
+            df = image_to_data(screenshot)
+            selection = df['text'] == key
+
+            # If there is no matching word, try again
+            if selection.sum() == 0:
+                logger.info('Found no match, trying again')
+                continue
+
+            # Exactly one word matching, exit the loop
+            elif selection.sum() == 1:
+                logger.info('Found exactly one match')
+                item = df[selection]
+                break
+
+            # More than one word matches, choose the first match
+            # Probably deterministic, but it should not be relied upon
+            else:
+                logger.info('Found multiple matches')
+                item = df[selection].iloc[0]
+                break
+
+        if item is None:
+            raise Exception(f"Found no key='{key}' in screenshots " +
+                            f"{first_scr} to {last_scr}")
+
+        x = float(item['left'] + item['width']/2)
+        y = float(item['top'] + item['height']/2)
+
+        self.mouse.move(x, y)
+        sleep(0.5)
+        self.mouse.click()
+        sleep(self.wait_time)
+
+    @action_decorator
+    @log_decorator
+    def kb_write(self, *args, **kwargs):
+        # delay is a workaround needed for keyboard library
+        kwargs.setdefault('delay', 0.1)
+        keyboard.write(*args, **kwargs)
+
+    @action_decorator
+    @log_decorator
+    def kb_send(self, *args, **kwargs):
+        keyboard.send(*args, **kwargs)
+
+    @log_decorator
+    def assert_text(self, key: str, timeout: float = 0):
+        """
+        Given key must be found in a screenshot before the timeout.
+
+        If the key is not found, exception is raised.
+        Zero timeout means that only one screenshot
+        will be taken and evaluated.
+        """
+
+        logger.info(f"Trying to find key='{key}'")
+
+        end_time = time() + timeout
+        first = True
+
+        while first or time() < end_time:
+            first = False
```

Comment:
Why do you use first? This loop is anyway ruled by time.

Reply:
This is needed for timeout=0, which should take only one screenshot. This is described in the docstring.